File: src/infrastructure/external/beads_ticket_reader.py
"""BeadsTicketReader -- anticorruption layer for the Ticket Freshness context.

Translates beads data into domain-layer value objects (OpenTicketData,
FreshnessFlag).  Reads ticket data from JSONL export and enriches with
labels and comments from the ``bd`` CLI (since JSONL lacks both).
Handles missing directories, missing CLI, and corrupted data gracefully.
"""

from __future__ import annotations

import json
import re
import subprocess
from typing import TYPE_CHECKING

from src.domain.models.ticket_freshness import (
    ContextDiff,
    FreshnessFlag,
    OpenTicketData,
)

if TYPE_CHECKING:
    from pathlib import Path

# Matches old-format ripple comments: **Ripple context diff from `<id>`:**
_RIPPLE_PATTERN_OLD = re.compile(
    r"\*\*Ripple context diff from `([^`]+)`:\*\*\s*(.*)",
    re.DOTALL,
)

# Matches current ripple comments: **Ripple review needed** -- `<id>` (title) was closed.
_RIPPLE_TRIGGER_RE = re.compile(
    r"\*\*Ripple review needed\*\*\s*--\s*`([^`]+)`",
)

# Extracts the "What changed:" summary text
_WHAT_CHANGED_RE = re.compile(
    r"\*\*What changed:\*\*\s*(.*?)(?:\n\n|\Z)",
    re.DOTALL,
)

# Parses bd query output lines: ○ alty-2j7.8 [● P2] [task] [review_needed] - title
_BD_QUERY_LINE_RE = re.compile(r"^[○◐●✓❄]\s+(\S+)")

_BD_TIMEOUT_SECONDS = 10


class BeadsTicketReader:
    """Reads beads data and translates it into domain value objects.

    This is an Anti-Corruption Layer (ACL) that shields the domain from
    beads data format details.  JSONL provides basic ticket data; the
    ``bd`` CLI provides labels and comments (which are not in the export).

    Attributes:
        _beads_dir: Path to the .beads directory.
    """

    def __init__(self, beads_dir: Path) -> None:
        self._beads_dir = beads_dir

    def read_open_tickets(self) -> tuple[OpenTicketData, ...]:
        """Read all open tickets from issues.jsonl, enriched with labels.

        Labels are fetched via ``bd query label=review_needed`` since the
        JSONL export does not include label data.

        Returns:
            Tuple of OpenTicketData for each open ticket.  Returns empty
            tuple if the directory or file is missing.
        """
        tickets = self._read_tickets_from_jsonl()
        flagged_ids = self._get_flagged_ids()
        return self._enrich_labels(tickets, flagged_ids)

    def read_flags(self, ticket_id: str) -> tuple[FreshnessFlag, ...]:
        """Read freshness flags from ripple review comments.

        Tries ``bd comments <id>`` first (primary source).  Falls back
        to interactions.jsonl if the CLI is unavailable.

        Args:
            ticket_id: The ticket to read flags for.

        Returns:
            Tuple of FreshnessFlag extracted from ripple comments.
        """
        flags = self._read_flags_from_bd_comments(ticket_id)
        if flags:
            return flags
        return self._read_flags_from_jsonl(ticket_id)

    # ------------------------------------------------------------------
    # JSONL reading (basic ticket data)
    # ------------------------------------------------------------------

    def _read_tickets_from_jsonl(self) -> list[OpenTicketData]:
        """Read open tickets from issues.jsonl (no labels)."""
        jsonl_path = self._beads_dir / "issues.jsonl"
        if not jsonl_path.exists():
            return []

        tickets: list[OpenTicketData] = []
        with jsonl_path.open() as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    data = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if data.get("status") != "open":
                    continue

                tickets.append(
                    OpenTicketData(
                        ticket_id=data.get("id", ""),
                        title=data.get("title", ""),
                        labels=(),
                        last_reviewed=None,
                    )
                )

        return tickets

    def _read_flags_from_jsonl(self, ticket_id: str) -> tuple[FreshnessFlag, ...]:
        """Fallback: read flags from interactions.jsonl."""
        interactions_path = self._beads_dir / "interactions.jsonl"
        if not interactions_path.exists():
            return ()

        flags: list[FreshnessFlag] = []
        with interactions_path.open() as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    data = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if data.get("issue_id") != ticket_id:
                    continue

                body = data.get("body", "")
                match = _RIPPLE_PATTERN_OLD.search(body)
                if not match:
                    continue

                triggering_id = match.group(1)
                summary = match.group(2).strip()
                if not summary:
                    continue

                created_at = data.get("created_at", "")
                flags.append(
                    FreshnessFlag(
                        context_diff=ContextDiff(
                            summary=summary,
                            triggering_ticket_id=triggering_id,
                            produced_at=created_at,
                        ),
                        flagged_at=created_at,
                    )
                )

        return tuple(flags)

    # ------------------------------------------------------------------
    # bd CLI integration (labels + comments)
    # ------------------------------------------------------------------

    def _get_flagged_ids(self) -> set[str]:
        """Get IDs of tickets with review_needed label via bd CLI."""
        try:
            result = subprocess.run(
                ["bd", "query", "label=review_needed"],  # noqa: S607
                capture_output=True,
                text=True,
                timeout=_BD_TIMEOUT_SECONDS,
            )
            if result.returncode != 0:
                return set()

            ids: set[str] = set()
            for line in result.stdout.splitlines():
                match = _BD_QUERY_LINE_RE.match(line.strip())
                if match:
                    ids.add(match.group(1))
            return ids
        except (FileNotFoundError, subprocess.TimeoutExpired):
            return set()

    @staticmethod
    def _enrich_labels(
        tickets: list[OpenTicketData],
        flagged_ids: set[str],
    ) -> tuple[OpenTicketData, ...]:
        """Replace empty labels with review_needed for flagged tickets."""
        enriched: list[OpenTicketData] = []
        for t in tickets:
            labels = ("review_needed",) if t.ticket_id in flagged_ids else ()
            enriched.append(
                OpenTicketData(
                    ticket_id=t.ticket_id,
                    title=t.title,
                    labels=labels,
                    last_reviewed=t.last_reviewed,
                )
            )
        return tuple(enriched)

    def _read_flags_from_bd_comments(self, ticket_id: str) -> tuple[FreshnessFlag, ...]:
        """Parse ripple review comments from bd comments output."""
        try:
            result = subprocess.run(  # noqa: S603
                ["bd", "comments", ticket_id],  # noqa: S607
                capture_output=True,
                text=True,
                timeout=_BD_TIMEOUT_SECONDS,
            )
            if result.returncode != 0:
                return ()
        except (FileNotFoundError, subprocess.TimeoutExpired):
            return ()

        return self._parse_bd_comments(result.stdout)

    @staticmethod
    def _parse_bd_comments(output: str) -> tuple[FreshnessFlag, ...]:
        """Extract ripple review flags from bd comments text output.

        Handles both current format (``**Ripple review needed** -- ...``)
        and old format (``**Ripple context diff from ...**``).
        """
        flags: list[FreshnessFlag] = []

        # Split into individual comments by the [Author] at <date> pattern
        comment_blocks = re.split(r"\n\[.+?\] at (\d{4}-\d{2}-\d{2}[\sT]?\d{0,8})", output)

        # comment_blocks[0] is header, then pairs of (date, body)
        i = 1
        while i < len(comment_blocks) - 1:
            date_str = comment_blocks[i].strip()
            body = comment_blocks[i + 1]
            i += 2

            # Try current format first
            trigger_match = _RIPPLE_TRIGGER_RE.search(body)
            changed_match = _WHAT_CHANGED_RE.search(body)

            if trigger_match and changed_match:
                triggering_id = trigger_match.group(1)
                summary = changed_match.group(1).strip()
                if summary:
                    flags.append(
                        FreshnessFlag(
                            context_diff=ContextDiff(
                                summary=summary,
                                triggering_ticket_id=triggering_id,
                                produced_at=date_str,
                            ),
                            flagged_at=date_str,
                        )
                    )
                continue

            # Try old format
            old_match = _RIPPLE_PATTERN_OLD.search(body)
            if old_match:
                triggering_id = old_match.group(1)
                summary = old_match.group(2).strip()
                if summary:
                    flags.append(
                        FreshnessFlag(
                            context_diff=ContextDiff(
                                summary=summary,
                                triggering_ticket_id=triggering_id,
                                produced_at=date_str,
                            ),
                            flagged_at=date_str,
                        )
                    )

        return tuple(flags)

File: internal/ticket/infrastructure/beads_ticket_reader.go
// Package infrastructure provides adapters for the Ticket bounded context.
package infrastructure

import (
	"bufio"
	"context"
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	ticketdomain "github.com/alty-cli/alty/internal/ticket/domain"
)

const bdTimeoutSeconds = 10

var (
	// Current format: **Ripple review needed** -- `<id>`
	rippleTriggerRE = regexp.MustCompile(`\*\*Ripple review needed\*\*\s*--\s*` + "`" + `([^` + "`" + `]+)` + "`")

	// Extracts "What changed:" summary text
	whatChangedRE = regexp.MustCompile(`\*\*What changed:\*\*\s*(.*?)(?:\n\n|\z)`)

	// Old format: **Ripple context diff from `<id>`:**
	ripplePatternOld = regexp.MustCompile(`\*\*Ripple context diff from ` + "`" + `([^` + "`" + `]+)` + "`" + `:\*\*\s*(.*)`)

	// Parses bd query output lines
	bdQueryLineRE = regexp.MustCompile(`^[○◐●✓❄]\s+(\S+)`)

	// Parses comment blocks
	commentBlockRE = regexp.MustCompile(`\n\[.+?\] at (\d{4}-\d{2}-\d{2}[\sT]?\d{0,8})`)
)

// BeadsTicketReader reads beads data and translates it into domain value objects.
// This is an Anti-Corruption Layer (ACL) that shields the domain from beads data format details.
type BeadsTicketReader struct {
	beadsDir string
}

// NewBeadsTicketReader creates a BeadsTicketReader.
func NewBeadsTicketReader(beadsDir string) *BeadsTicketReader {
	return &BeadsTicketReader{beadsDir: beadsDir}
}

// ReadOpenTickets reads all open tickets from issues.jsonl, enriched with labels.
func (r *BeadsTicketReader) ReadOpenTickets(ctx context.Context) []ticketdomain.OpenTicketData {
	tickets := r.readTicketsFromJSONL()
	flaggedIDs := r.getFlaggedIDs(ctx)
	return r.enrichLabels(tickets, flaggedIDs)
}

// ReadFlags reads freshness flags from ripple review comments.
func (r *BeadsTicketReader) ReadFlags(ctx context.Context, ticketID string) []ticketdomain.FreshnessFlag {
	flags := r.readFlagsFromBDComments(ctx, ticketID)
	if len(flags) > 0 {
		return flags
	}
	return r.readFlagsFromJSONL(ticketID)
}

// ------------------------------------------------------------------
// JSONL reading
// ------------------------------------------------------------------

type issueJSON struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Status string `json:"status"`
}

func (r *BeadsTicketReader) readTicketsFromJSONL() []ticketdomain.OpenTicketData {
	jsonlPath := filepath.Join(r.beadsDir, "issues.jsonl")
	f, err := os.Open(jsonlPath)
	if err != nil {
		return nil
	}
	defer func() { _ = f.Close() }()

	var tickets []ticketdomain.OpenTicketData
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var issue issueJSON
		if err := json.Unmarshal([]byte(line), &issue); err != nil {
			continue
		}
		if issue.Status != "open" {
			continue
		}
		tickets = append(tickets, ticketdomain.NewOpenTicketData(issue.ID, issue.Title, nil, nil))
	}
	return tickets
}

type interactionJSON struct {
	IssueID   string `json:"issue_id"`
	Body      string `json:"body"`
	CreatedAt string `json:"created_at"`
}

func (r *BeadsTicketReader) readFlagsFromJSONL(ticketID string) []ticketdomain.FreshnessFlag {
	interactionsPath := filepath.Join(r.beadsDir, "interactions.jsonl")
	f, err := os.Open(interactionsPath)
	if err != nil {
		return nil
	}
	defer func() { _ = f.Close() }()

	var flags []ticketdomain.FreshnessFlag
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var interaction interactionJSON
		if err := json.Unmarshal([]byte(line), &interaction); err != nil {
			continue
		}
		if interaction.IssueID != ticketID {
			continue
		}

		match := ripplePatternOld.FindStringSubmatch(interaction.Body)
		if match == nil {
			continue
		}
		triggeringID := match[1]
		summary := strings.TrimSpace(match[2])
		if summary == "" {
			continue
		}

		diff, err := ticketdomain.NewContextDiff(summary, triggeringID, interaction.CreatedAt)
		if err != nil {
			continue
		}
		flags = append(flags, ticketdomain.NewFreshnessFlag(diff, interaction.CreatedAt))
	}
	return flags
}

// ------------------------------------------------------------------
// bd CLI integration
// ------------------------------------------------------------------

func (r *BeadsTicketReader) getFlaggedIDs(ctx context.Context) map[string]struct{} {
	ctx, cancel := context.WithTimeout(ctx, bdTimeoutSeconds*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "bd", "query", "label=review_needed")
	output, err := cmd.Output()
	if err != nil {
		return nil
	}

	ids := make(map[string]struct{})
	for _, line := range strings.Split(string(output), "\n") {
		match := bdQueryLineRE.FindStringSubmatch(strings.TrimSpace(line))
		if match != nil {
			ids[match[1]] = struct{}{}
		}
	}
	return ids
}

func (r *BeadsTicketReader) enrichLabels(tickets []ticketdomain.OpenTicketData, flaggedIDs map[string]struct{}) []ticketdomain.OpenTicketData {
	enriched := make([]ticketdomain.OpenTicketData, len(tickets))
	for i, t := range tickets {
		var labels []string
		if _, ok := flaggedIDs[t.TicketID()]; ok {
			labels = []string{"review_needed"}
		}
		enriched[i] = ticketdomain.NewOpenTicketData(t.TicketID(), t.Title(), labels, t.LastReviewed())
	}
	return enriched
}

func (r *BeadsTicketReader) readFlagsFromBDComments(ctx context.Context, ticketID string) []ticketdomain.FreshnessFlag {
	ctx, cancel := context.WithTimeout(ctx, bdTimeoutSeconds*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "bd", "comments", ticketID)
	output, err := cmd.Output()
	if err != nil {
		return nil
	}
	return parseBDComments(string(output))
}

func parseBDComments(output string) []ticketdomain.FreshnessFlag {
	var flags []ticketdomain.FreshnessFlag

	blocks := commentBlockRE.Split(output, -1)
	dates := commentBlockRE.FindAllStringSubmatch(output, -1)

	for i, dateMatch := range dates {
		if i+1 >= len(blocks) {
			break
		}
		dateStr := strings.TrimSpace(dateMatch[1])
		body := blocks[i+1]

		// Try current format
		triggerMatch := rippleTriggerRE.FindStringSubmatch(body)
		changedMatch := whatChangedRE.FindStringSubmatch(body)
		if triggerMatch != nil && changedMatch != nil {
			summary := strings.TrimSpace(changedMatch[1])
			if summary != "" {
				diff, err := ticketdomain.NewContextDiff(summary, triggerMatch[1], dateStr)
				if err == nil {
					flags = append(flags, ticketdomain.NewFreshnessFlag(diff, dateStr))
				}
			}
			continue
		}

		// Try old format
		oldMatch := ripplePatternOld.FindStringSubmatch(body)
		if oldMatch != nil {
			summary := strings.TrimSpace(oldMatch[2])
			if summary != "" {
				diff, err := ticketdomain.NewContextDiff(summary, oldMatch[1], dateStr)
				if err == nil {
					flags = append(flags, ticketdomain.NewFreshnessFlag(diff, dateStr))
				}
			}
		}
	}
	return flags
}
